File: Rohit Singhal/Recursion+Backtracking+DP/LongestCommonSubsequence.java
class Solution {
    
    public int helper(String s1,String s2, int i, int j,int[][] dp)
    {
        if(i>=s1.length() || j>=s2.length())
        {
            return 0;
        }
        
        if(dp[i][j]!=-1)
        {
            return dp[i][j];
        }
        if(s1.charAt(i)==s2.charAt(j))
        {
            dp[i][j]=1+helper(s1,s2,i+1,j+1,dp);
            return dp[i][j];
        }
        else
        {
            dp[i][j]= Math.max(helper(s1,s2,i+1,j,dp),helper(s1,s2,i,j+1,dp));
            return dp[i][j];
        }
    }
    
    
    
    public int longestCommonSubsequence(String text1, String text2) {
        
        int[][] dp=new int[text1.length()][text2.length()];
        for(int i=0;i<dp.length;i++)
        {
            Arrays.fill(dp[i],-1);
        }
        return helper(text1,text2,0,0,dp);
        
    }
}

File: Shreya/Recursion and DP/Longest Common Subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        if(text1.length()==0 || text2.length()==0)
        {
            return 0;
        }
        vector<vector<int>>vec(text1.length()+1,vector<int>(text2.length()+1,0));
        for(int i=0;i<text1.length();i++)
        {
            for(int j=0;j<text2.length();j++)
            {
                if (text1[i]==text2[j])
                {
                    vec[i+1][j+1]=1+vec[i][j];
                }
                else
                {
                    vec[i+1][j+1]=max(vec[i][j+1],vec[i+1][j]);
                }
            }
        }
        
       return vec[text1.length()][text2.length()]; 
    }
};